syncer/syncer.go

```diff
@@ -467,15 +464,39 @@ func (s *Syncer) IsFreshTask() (bool, error) {
 }
 
 func (s *Syncer) resetReplicationSyncer() {
```

Comment:
what about
```golang
func (s *Syncer) resetReplicationSyncer() {
    // close streamer producer
    if s.streamerProducer != nil {
        switch t := s.streamerProducer.(type) {
        case *remoteBinlogReader:
            s.closeBinlogSyncer(t.reader)
        case *localBinlogReader:
            // TODO: close old local reader before creating a new one
        }
    }
    // re-create streamer producer
    switch s.binlogType {
    case RemoteBinlog:
        s.streamerProducer = &remoteBinlogReader{replication.NewBinlogSyncer(s.syncCfg), s.tctx, s.cfg.EnableGTID}
    case LocalBinlog:
        s.streamerProducer = &localBinlogReader{streamer.NewBinlogReader(s.tctx, &streamer.BinlogReaderConfig{RelayDir: s.cfg.RelayDir, Timezone: s.timezone})}
    default:
        s.tctx.L().Error("init streamerProducer with un-recognized binlogType")
    }
}
```

Reply:
It's intentional, for some tests, if syncer already has a mock streamerProducer, it won't re-create new streamerProducer otherwise we need add mock streamerProducer type here.

Reply:
well I see, but it is implicit, we'd better add some comment.

```golang
func (s *Syncer) resetReplicationSyncer() {
    // close streamer producer
    if s.streamerProducer != nil {
        switch t := s.streamerProducer.(type) {
        case *remoteBinlogReader:
            s.closeBinlogSyncer(t.reader)
        case *localBinlogReader:
            // TODO: close old local reader before creating a new one
        default:
            // NOTE: add some comment for mock streamer producer here?
            return
        }
    }
    // re-create streamer producer
    switch s.binlogType {
    case RemoteBinlog:
        s.streamerProducer = &remoteBinlogReader{replication.NewBinlogSyncer(s.syncCfg), s.tctx, s.cfg.EnableGTID}
    case LocalBinlog:
        s.streamerProducer = &localBinlogReader{streamer.NewBinlogReader(s.tctx, &streamer.BinlogReaderConfig{RelayDir: s.cfg.RelayDir, Timezone: s.timezone})}
    default:
        s.tctx.L().Error("init streamerProducer with un-recognized binlogType")
    }
}
```

Reply:
Done in https://github.com/pingcap/dm/pull/226/files#diff-ea4abf20044394891136e70431919a4fR475